File: python/balanced_binary_tree.py
import sys

sys.path.append("./models")
sys.path.append("./utils")

from typing import Optional
from binary_tree_node import TreeNode
from deserialize_to_binary_tree import DeserializeFromList
from serialize_binary_tree import SerializeBinaryTreeToList


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        """time complexity: O(n)"""
        return self.dfs(root)[0]

    def dfs(self, root: Optional[TreeNode]) -> tuple[bool, int]:
        """postorder traversal"""
        if root is None:
            return (True, 0)
        left: tuple[bool, int] = self.dfs(root.left)
        right: tuple[bool, int] = self.dfs(root.right)
        is_balanced: bool = left[0] and right[0] and abs(left[1] - right[1]) <= 1
        return (is_balanced, max(left[1], right[1]) + 1)


def test_isBalanced_case_1():
    # arrange
    root: Optional[TreeNode] = DeserializeFromList([3, 9, 20, None, None, 15, 7])
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.isBalanced(root)

    # assert
    assert expected == actual


def test_isBalanced_case_2():
    # arrange
    arr: list[Optional[int]] = [1, 2, 2, 3, 3, None, None, 4, 4]
    root: Optional[TreeNode] = DeserializeFromList(arr)
    expected: bool = False

    # act
    solution = Solution()
    actual = solution.isBalanced(root)

    # assert
    assert expected == actual


def test_isBalanced_case_3():
    # arrange
    root: Optional[TreeNode] = None
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.isBalanced(root)

    # assert
    assert expected == actual

File: rust/src/code/balanced_binary_tree.rs
use super::models::binary_tree_node::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

struct Solution {}

impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::helper(&root).0
    }

    fn helper(root: &Option<Rc<RefCell<TreeNode>>>) -> (bool, i32) {
        if let Some(node) = root {
            let left = Self::helper(&node.borrow().left);
            let right = Self::helper(&node.borrow().right);
            let balanced = left.0 && right.0 && (left.1 - right.1).abs() <= 1;

            (balanced, std::cmp::max(left.1, right.1) + 1)
        } else {
            (true, 0)
        }
    }
}

#[cfg(test)]
mod test {
    use super::Solution;
    use crate::code::utils::deserialize_to_binary_tree::deserialize_to_BT;

    #[test]
    fn balanced_binary_tree_case_1() {
        // arrange
        let root = vec![Some(3), Some(9), Some(20), None, None, Some(15), Some(7)];
        let expected = true;

        // act
        let actual = Solution::is_balanced(deserialize_to_BT(root));

        // assert
        assert_eq!(expected, actual)
    }

    #[test]
    fn balanced_binary_tree_case_2() {
        // arrange
        let root = vec![
            Some(1),
            Some(2),
            Some(2),
            Some(3),
            Some(3),
            None,
            None,
            Some(4),
            Some(4),
        ];
        let expected = false;

        // act
        let actual = Solution::is_balanced(deserialize_to_BT(root));

        // assert
        assert_eq!(expected, actual)
    }

    #[test]
    fn balanced_binary_tree_case_3() {
        // arrange
        let root = vec![];
        let expected = true;

        // act
        let actual = Solution::is_balanced(deserialize_to_BT(root));

        // assert
        assert_eq!(expected, actual)
    }
}
